### PR TITLE
add shoppingCart/ cartItem tests

### PR DESCRIPTION
**Pull request by mmosii**
https://paste.pics/72c6829a2fc7a34ac29d785eac33a9b4

### CODE REVIEW

File: src/test/resources/database/add-shopping-cart-and-user.sql [hunk 1]
```diff
@@ -0,0 +1,3 @@
+insert into users values (5,'test@gmail.com', 'testpass', 'Brad', 'Pitt', 'home', false);
+insert into shopping_carts values (3, 5);
+insert into cart_items values (6, 3, 2, 10);
```
Comment (RedArtelerist) [inline src/test/resources/database/add-shopping-cart-and-user.sql]: I think it's better to add column names for clear understanding of what you insert. Think if you open this file on your job you will need to open tables in db and it's not comfortable.

File: src/test/java/mmosii/bookstore/controller/ShoppingCartControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                .webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+    }
+
+    @Test
+    @DisplayName("Get shopping cart for valid user")
+    @WithMockUser(username = "test@gmail.com", authorities = {"USER", "ADMIN"})
+    void getShoppingCart_validUser_returnsDto() throws Exception {
+        ShoppingCartDto expected = new ShoppingCartDto(3L, 5L, null);
```
Comment (RedArtelerist) [inline src/test/java/mmosii/bookstore/controller/ShoppingCartControllerTest.java]: Why null, why not the empty list?

File: src/test/java/mmosii/bookstore/controller/ShoppingCartControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        MvcResult result = mockMvc.perform(get("/api/cart")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        ShoppingCartDto actual = objectMapper.readValue(result.getResponse()
+                .getContentAsString(), ShoppingCartDto.class);
+        assertThat(actual).isNotNull();
+        assertThat(actual.id()).isEqualTo(expected.id());
+        assertThat(actual.userId()).isEqualTo(expected.userId());
```
Comment (RedArtelerist) [inline src/test/java/mmosii/bookstore/controller/ShoppingCartControllerTest.java]: You can inline it using one assertThat 
```suggestion
        assertThat(actual).isNotNull()
        .hasFieldOrPropertyWithValue("id", expected.id())
       .hasFieldOrPropertyWithValue("userId", expected.userId());
```
Something like this

File: src/test/java/mmosii/bookstore/controller/ShoppingCartControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    @DisplayName("Add valid cartItem to shopping cart")
+    @WithMockUser(username = "test@gmail.com", authorities = {"USER", "ADMIN"})
+    void addBookToCart_validRequestDto_returnsDto() throws Exception {
+        CartItemRequestDto requestDto = new CartItemRequestDto(2L, 10);
+        CartItemDto expected = new CartItemDto(null, 2L, "Shantaram", 10);
+
+        String jsonRequest = objectMapper.writeValueAsString(requestDto);
+        MvcResult result = mockMvc.perform(post("/api/cart")
+                        .content(jsonRequest)
+                        .contentType(MediaType.APPLICATION_JSON))
```
Comment (RedArtelerist) [inline src/test/java/mmosii/bookstore/controller/ShoppingCartControllerTest.java]:
```suggestion
        MvcResult result = mockMvc.perform(post("/api/cart")
                        .content(objectMapper.writeValueAsString(requestDto))
                        .contentType(MediaType.APPLICATION_JSON))
```

File: src/test/java/mmosii/bookstore/repository/CartItemRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        assertThat(actual.getQuantity()).isEqualTo(10);
+        assertThat(actual.getBook().getTitle()).isEqualTo("Shantaram");
+    }
+
+    @Test
+    @DisplayName("Find shopping cart by non existing user id")
+    void findByIdAndShoppingCartId_notValidIds_returnsOptionalEmpty() {
+        Optional<CartItem> actual = cartItemRepository
+                .findByIdAndShoppingCartId(50L, 50L);
+        assertThat(actual).isEqualTo(Optional.empty());
```
Comment (RedArtelerist) [inline src/test/java/mmosii/bookstore/repository/CartItemRepositoryTest.java]: If I understand they are not custom repository methods (without @Query), so what is the reason for testing them? But you can leave it as is

File: src/test/java/mmosii/bookstore/service/ShoppingCartServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextHolder;
+
+@ExtendWith(MockitoExtension.class)
+public class ShoppingCartServiceTest {
+    @Mock
+    private ShoppingCartRepository shoppingCartRepository;
+    @Mock
+    private ShoppingCartMapper shoppingCartMapper;
```
Comment (RedArtelerist) [inline src/test/java/mmosii/bookstore/service/ShoppingCartServiceTest.java]: You can try don't mock mappers and make them @Spy. 
```
@Spy
private ShoppingCartMapper shoppingCartMapper = new ShoppingCartMapper Impl();
```
It decreases your amount of code.

File: src/test/java/mmosii/bookstore/service/ShoppingCartServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+        assertThat(actual.userId()).isEqualTo(expected.userId());
+        verify(securityContext, times(1)).getAuthentication();
+        verify(authentication, times(1)).getName();
+        verify(userRepository, times(1)).findByEmail("test@gmail.com");
+        verify(shoppingCartRepository, times(1)).findByUserId(user.getId());
+        verify(shoppingCartRepository, times(1)).save(any());
+        verify(shoppingCartMapper, times(1)).toDto(shoppingCart);
+        verifyNoMoreInteractions(securityContext, authentication,
+                userRepository, shoppingCartRepository, shoppingCartMapper);
```
Comment (RedArtelerist) [inline src/test/java/mmosii/bookstore/service/ShoppingCartServiceTest.java]: From mockito documentation:

> A word of warning: Some users who did a lot of classic, expect-run-verify mocking tend to use verifyNoMoreInteractions() very often, even in every test method. verifyNoMoreInteractions() is not recommended to use in every test method. verifyNoMoreInteractions() is a handy assertion from the interaction testing toolkit. **Use it only when it's relevant. Abusing it leads to over-specified, less maintainable tests.**

And also about verify:

> **Although it is possible to verify a stubbed invocation, usually it's just redundant.** Let's say you've stubbed foo.bar(). If your code cares what foo.bar() returns then something else breaks(often before even verify() gets executed). If your code doesn't care what foo.bar() returns then it should not be stubbed.

I think this methods more useful for testing void methods. So you can skip this comment or think to delete them (from Nazar's words) :)

File: src/test/java/mmosii/bookstore/service/ShoppingCartServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+    private ShoppingCartServiceImpl shoppingCartService;
+    @Mock
+    private Authentication authentication;
+    @Mock
+    private SecurityContext securityContext;
+
+    @Test
+    @DisplayName("Get already existing shopping cart")
+    void getShoppingCart_existingCart_returnsCartDto() {
+        SecurityContextHolder.setContext(securityContext);
```
Comment (RedArtelerist) [inline src/test/java/mmosii/bookstore/service/ShoppingCartServiceTest.java]: Do you really need to do it in every test method? Maybe it works without it? But if yes you can use it in BeforeEach

**Review by RedArtelerist — CHANGES_REQUESTED** [review]
Good job, I left some comments to think

File: src/test/java/mmosii/bookstore/controller/ShoppingCartControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+import org.springframework.web.context.WebApplicationContext;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@Sql(scripts = {"classpath:database/add-shopping-cart-and-user.sql",
+        "classpath:database/add-books-and-categories-to-books-table.sql"})
+@Sql(scripts = {"classpath:database/delete-shoppingcarts-and-user.sql",
+        "classpath:database/delete-books-and-categories-from-books-table.sql"},
+        executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+public class ShoppingCartControllerTest {
+    protected static MockMvc mockMvc;
```
Comment (NazarSmith228) [inline src/test/java/mmosii/bookstore/controller/ShoppingCartControllerTest.java]: why protected?

File: src/test/java/mmosii/bookstore/controller/ShoppingCartControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        MvcResult result = mockMvc.perform(get("/api/cart")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        ShoppingCartDto actual = objectMapper.readValue(result.getResponse()
+                .getContentAsString(), ShoppingCartDto.class);
+        assertThat(actual).isNotNull()
+                .hasFieldOrPropertyWithValue("id", expected.id())
+                .hasFieldOrPropertyWithValue("userId", expected.userId());
```
Comment (NazarSmith228) [inline src/test/java/mmosii/bookstore/controller/ShoppingCartControllerTest.java]:
```suggestion
                .hasFieldOrPropertyWithValue("id", expected.id())
                .hasFieldOrPropertyWithValue("userId", expected.userId())
                .hasFieldOrPropertyWithValue("list", Collections.emptyList());
```

File: src/test/java/mmosii/bookstore/controller/ShoppingCartControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    @Test
+    @DisplayName("Get shopping cart for valid user")
+    @WithMockUser(username = "test@gmail.com", authorities = {"USER", "ADMIN"})
+    void getShoppingCart_validUser_returnsDto() throws Exception {
+        ShoppingCartDto expected = new ShoppingCartDto(3L, 5L, Collections.emptyList());
+
+        MvcResult result = mockMvc.perform(get("/api/cart")
+                        .contentType(MediaType.APPLICATION_JSON))
```
Comment (NazarSmith228) [inline src/test/java/mmosii/bookstore/controller/ShoppingCartControllerTest.java]: this is a GET - why do you need content type?

File: src/test/java/mmosii/bookstore/controller/ShoppingCartControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    @DisplayName("Delete cartItem from shopping cart")
+    @WithMockUser(username = "test@gmail.com", authorities = {"USER", "ADMIN"})
+    void deleteBookFromCart_validId() throws Exception {
+        Long id = 6L;
+
+        mockMvc.perform(delete("/api/cart/cart-items/" + id)
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isNoContent())
+                .andReturn();
+    }
```
Comment (NazarSmith228) [inline src/test/java/mmosii/bookstore/controller/ShoppingCartControllerTest.java:130]: makes sense to verify that cart item with id 6 is no longer present in DB

File: src/test/java/mmosii/bookstore/repository/CartItemRepositoryTest.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+import mmosii.bookstore.repository.shoppingcart.CartItemRepository;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.test.context.jdbc.Sql;
+
+@DataJpaTest
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
```
Comment (NazarSmith228) [inline src/test/java/mmosii/bookstore/repository/CartItemRepositoryTest.java]: do you really need this? `NONE` means use autoconfigured datasource, which is H2 here

therefore I guess you can skip this annotation - pls verify

**Applicable to all such tests**

File: src/test/java/mmosii/bookstore/repository/CartItemRepositoryTest.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    @Autowired
+    private CartItemRepository cartItemRepository;
+
+    @Test
+    @DisplayName("Find shopping cart by valid user id")
+    void findByIdAndShoppingCartId_validIds_returnsCartItem() {
+        CartItem actual = cartItemRepository
+                .findByIdAndShoppingCartId(6L, 3L).get();
+        assertThat(actual.getQuantity()).isEqualTo(10);
+        assertThat(actual.getBook().getTitle()).isEqualTo("Shantaram");
```
Comment (NazarSmith228) [inline src/test/java/mmosii/bookstore/repository/CartItemRepositoryTest.java]:
```suggestion
        assertThat(cartItemRepository.findByIdAndShoppingCartId(6L, 3L))
            .get()
            .hasFieldOrPropertyWithValue("quantity", 10)
            .extracting(CartItem::getBook)
            .extracting(Book::getTitle)
            .isEqualTo("Shantaram");
```

File: src/test/java/mmosii/bookstore/repository/CartItemRepositoryTest.java (hunk 5 above)
Comment (NazarSmith228) [inline src/test/java/mmosii/bookstore/repository/CartItemRepositoryTest.java]:
```suggestion
        assertThat(cartItemRepository.findByIdAndShoppingCartId(50L, 50L)).isEmpty();
```

File: src/test/java/mmosii/bookstore/repository/ShoppingCartRepositoryTest.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+        executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+public class ShoppingCartRepositoryTest {
+    @Autowired
+    private ShoppingCartRepository shoppingCartRepository;
+
+    @Test
+    @DisplayName("Find shopping cart by valid user id")
+    void findByUserId_validId_returnsShoppingCart() {
+        ShoppingCart actual = shoppingCartRepository.findByUserId(5L).get();
+        assertThat(actual.getId()).isEqualTo(3L);
```
Comment (NazarSmith228) [inline src/test/java/mmosii/bookstore/repository/ShoppingCartRepositoryTest.java]:
```suggestion
        assertThat(shoppingCartRepository.findByUserId(5L))
              .get()
              // assertions on fields
```

File: src/test/java/mmosii/bookstore/repository/ShoppingCartRepositoryTest.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    @Test
+    @DisplayName("Find shopping cart by non existing user id")
+    void findByUserId_notValidId_returnsOptionalEmpty() {
+        Optional<ShoppingCart> actual = shoppingCartRepository.findByUserId(50L);
+        assertThat(actual).isEqualTo(Optional.empty());
```
Comment (NazarSmith228) [inline src/test/java/mmosii/bookstore/repository/ShoppingCartRepositoryTest.java]:
```suggestion
        assertThat(shoppingCartRepository.findByUserId(50L)).isEmpty();
```

File: src/test/java/mmosii/bookstore/service/ShoppingCartServiceTest.java (hunk 6 above)
Comment (NazarSmith228) [reply src/test/java/mmosii/bookstore/service/ShoppingCartServiceTest.java]: @mmosii pls address Maksym's comment - `@Spy` will reduce the amount of boilerplate in your tests

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

**Review by NazarSmith228 — APPROVED** [review]
seems LGTM, but pls resolve the merge conflicts

**Pull request merged**